# Feature Engineering

---

1. Import packages
2. Load data
3. Feature engineering

---

## 1. Import packages

In [1]:
# Data manipulation
import numpy as np
import pandas as pd
import datetime as dt

# Data visualization
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

# Shows plots in jupyter notebook
%matplotlib inline

# Set plot style
sns.set(color_codes=True)

# Ignore warning
import warnings
warnings.filterwarnings('ignore')

In [2]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [3]:
print('NumPy', np.__version__)
print('Pandas', pd.__version__)
print('Matplotlib', mpl.__version__)
print('Seaborn', sns.__version__)

NumPy 1.23.5
Pandas 2.0.1
Matplotlib 3.7.1
Seaborn 0.12.2


---
## 2. Load data

In [4]:
# Read the data
df = pd.read_csv('./clean_data_after_eda.csv')

# # Convert date columns to datetime
# df['date_activ'] = pd.to_datetime(df['date_activ'], format='%Y-%m-%d')
# df['date_end'] = pd.to_datetime(df['date_end'], format='%Y-%m-%d')
# df['date_modif_prod'] = pd.to_datetime(df['date_modif_prod'], format='%Y-%m-%d')
# df['date_renewal'] = pd.to_datetime(df['date_renewal'], format='%Y-%m-%d')
date_columns = ['date_activ', 'date_end', 'date_modif_prod', 'date_renewal']
for col in date_columns:
    df[col] = pd.to_datetime(df[col], format='%Y-%m-%d')

In [5]:
df.head()

Unnamed: 0                                id   
0           0  24011ae4ebbe3035111d65fa7c15bc57  \
1           1  d29c2c54acc38ff3c0614d0a653813dd   
2           2  764c75f661154dac3a6c254cd082ea7d   
3           3  bba03439a292a1e166f80264c16191cb   
4           4  149d57cf92fc41cf94415803a877cb4b   

                      channel_sales  cons_12m  cons_gas_12m  cons_last_month   
0  foosdfpfkusacimwkcsosbicdxkicaua         0         54946                0  \
1                           MISSING      4660             0                0   
2  foosdfpfkusacimwkcsosbicdxkicaua       544             0                0   
3  lmkebamcaaclubfxadlmueccxoimlema      1584             0                0   
4                           MISSING      4425             0              526   

  date_activ   date_end date_modif_prod date_renewal  forecast_cons_12m   
0 2013-06-15 2016-06-15      2015-11-01   2015-06-23               0.00  \
1 2009-08-21 2016-08-30      2009-08-21   2015-08-31             189.95   
2 2010-04-16 2016-04-16      2010-04-16   2015-04-17              47.96   
3 2010-03-30 2016-03-30      2010-03-30   2015-03-31             240.04   
4 2010-01-13 2016-03-07      2010-01-13   2015-03-09             445.75   

   forecast_cons_year  forecast_discount_energy  forecast_meter_rent_12m   
0                   0                       0.0                     1.78  \
1                   0                       0.0                    16.27   
2                   0                       0.0                    38.72   
3                   0                       0.0                    19.83   
4                 526                       0.0                   131.73   

   forecast_price_energy_off_peak  forecast_price_energy_peak   
0                        0.114481                    0.098142  \
1                        0.145711                    0.000000   
2                        0.165794                    0.087899   
3                        0.146694                    0.000000   
4                        0.116900                    0.100015   

   forecast_price_pow_off_peak has_gas  imp_cons  margin_gross_pow_ele   
0                    40.606701       t      0.00                 25.44  \
1                    44.311378       f      0.00                 16.38   
2                    44.311378       f      0.00                 28.60   
3                    44.311378       f      0.00                 30.22   
4                    40.606701       f     52.32                 44.91   

   margin_net_pow_ele  nb_prod_act  net_margin  num_years_antig   
0               25.44            2      678.99                3  \
1               16.38            1       18.89                6   
2               28.60            1        6.60                6   
3               30.22            1       25.46                6   
4               44.91            1       47.98                6   

                          origin_up  pow_max                   price_date_x   
0  lxidpiddsbxsbosboudacockeimpuepw   43.648  2015-06-16 12:00:00.000000000  \
1  kamkkxfxxuwbdslkwifmmcsiusiuosws   13.800  2015-06-16 12:00:00.000000000   
2  kamkkxfxxuwbdslkwifmmcsiusiuosws   13.856  2015-06-16 12:00:00.000000000   
3  kamkkxfxxuwbdslkwifmmcsiusiuosws   13.200  2015-06-16 12:00:00.000000000   
4  kamkkxfxxuwbdslkwifmmcsiusiuosws   19.800  2015-06-16 12:00:00.000000000   

   mean_year_price_p1_var  mean_year_price_p2_var  mean_year_price_p3_var   
0                0.124787                0.100749                0.066530  \
1                0.149609                0.007124                0.000000   
2                0.170512                0.088421                0.000000   
3                0.151210                0.000000                0.000000   
4                0.124174                0.103638                0.072865   

   mean_year_price_p1_fix  mean_year_price_p2_fix  mean_year_price_p3_fix   
0               40.942265               22.352010    

In [6]:
df = df.drop(['Unnamed: 0'], axis=1)

---

## 3. Feature engineering

### Difference between off-peak prices in December and preceding January

Below is the code created by our colleague to calculate the feature described above. Use this code to re-create this feature and then think about ways to build on this feature to create features with a higher predictive power.

In [7]:
# Read price data
price_df = pd.read_csv('price_data.csv')
price_df['price_date'] = pd.to_datetime(price_df['price_date'], format='%Y-%m-%d')

In [8]:
price_df.head()

id price_date  price_off_peak_var   
0  038af19179925da21a25619c5a24b745 2015-01-01            0.151367  \
1  038af19179925da21a25619c5a24b745 2015-02-01            0.151367   
2  038af19179925da21a25619c5a24b745 2015-03-01            0.151367   
3  038af19179925da21a25619c5a24b745 2015-04-01            0.149626   
4  038af19179925da21a25619c5a24b745 2015-05-01            0.149626   

   price_peak_var  price_mid_peak_var  price_off_peak_fix  price_peak_fix   
0             0.0                 0.0           44.266931             0.0  \
1             0.0                 0.0           44.266931             0.0   
2             0.0                 0.0           44.266931             0.0   
3             0.0                 0.0           44.266931             0.0   
4             0.0                 0.0           44.266931             0.0   

   price_mid_peak_fix  
0                 0.0  
1                 0.0  
2                 0.0  
3                 0.0  
4                 0.0

We first need to recreate the feature that our colleague has already created, which is "the difference between off-peak prices in December and January the preceding year". We can do this by grouping the off-peak prices by companies and month, and then calculating the difference between the prices in December and January.

In [9]:
# Group off-peak prices by companies and month
monthly_price_by_id = price_df.groupby(['id', 'price_date']).agg({'price_off_peak_var': 'mean', 'price_off_peak_fix': 'mean'}).reset_index()

# Get January and December prices
jan_prices = monthly_price_by_id.groupby('id').first().reset_index()
dec_prices = monthly_price_by_id.groupby('id').last().reset_index()

# Calculate the difference
diff = pd.merge(dec_prices.rename(columns={'price_off_peak_var': 'dec_1', 'price_off_peak_fix': 'dec_2'}), jan_prices.drop(columns='price_date'), on='id')
diff['offpeak_diff_dec_jan_energy'] = diff['dec_1'] - diff['price_off_peak_var']
diff['offpeak_diff_dec_jan_power'] = diff['dec_2'] - diff['price_off_peak_fix']
diff = diff[['id', 'offpeak_diff_dec_jan_energy', 'offpeak_diff_dec_jan_power']]

In [10]:
diff.head()

id  offpeak_diff_dec_jan_energy   
0  0002203ffbb812588b632b9e628cc38d                    -0.006192  \
1  0004351ebdd665e6ee664792efc4fd13                    -0.004104   
2  0010bcc39e42b3c2131ed2ce55246e3c                     0.050443   
3  0010ee3855fdea87602a5b7aba8e42de                    -0.010018   
4  00114d74e963e47177db89bc70108537                    -0.003994   

   offpeak_diff_dec_jan_power  
0                    0.162916  
1                    0.177779  
2                    1.500000  
3                    0.162916  
4                   -0.000001

In [11]:
# Group off-peak prices by companies and month for other months
other_month_prices = monthly_price_by_id.groupby('id').agg({'price_off_peak_var': 'mean', 'price_off_peak_fix': 'mean'}).reset_index()

# Calculate the difference for other months
diff_other_months = pd.merge(other_month_prices.rename(columns={'price_off_peak_var': 'other_month_1', 'price_off_peak_fix': 'other_month_2'}), jan_prices.drop(columns='price_date'), on='id')
diff_other_months['offpeak_diff_other_months_energy'] = diff_other_months['other_month_1'] - diff_other_months['price_off_peak_var']
diff_other_months['offpeak_diff_other_months_power'] = diff_other_months['other_month_2'] - diff_other_months['price_off_peak_fix']
diff_other_months = diff_other_months[['id', 'offpeak_diff_other_months_energy', 'offpeak_diff_other_months_power']]

In [12]:
diff_other_months.head()

id  offpeak_diff_other_months_energy   
0  0002203ffbb812588b632b9e628cc38d                         -0.001760  \
1  0004351ebdd665e6ee664792efc4fd13                         -0.001621   
2  0010bcc39e42b3c2131ed2ce55246e3c                          0.030721   
3  0010ee3855fdea87602a5b7aba8e42de                         -0.004329   
4  00114d74e963e47177db89bc70108537                         -0.001508   

   offpeak_diff_other_months_power  
0                     1.357630e-01  
1                     1.185192e-01  
2                     8.750000e-01  
3                     8.145810e-02  
4                    -8.000000e-07

In [13]:
# Replace 't' with 1 and 'f' with 0 in the 'has_gas' column
df['has_gas'] = df['has_gas'].replace({'t': 1, 'f': 0})
df['has_gas'].unique()

array([1, 0], dtype=int64)

In [14]:
# Get a list of columns with object data type
object_features = df.select_dtypes(include=['object', 'datetime64']).columns.tolist()

# Display object features
print('Object Features:')
print(object_features)

Object Features:
['id', 'channel_sales', 'date_activ', 'date_end', 'date_modif_prod', 'date_renewal', 'origin_up', 'price_date_x', 'price_date_y', 'price_date']


In [15]:
# Add 'churn' to the list of columns to be dropped
columns_to_drop = object_features + ['churn']

---

## 4. Feature importances

Before training the Random Forest classifier, we should perform feature selection to choose the most relevant features for our model. We can use the feature importance provided by the Random Forest classifier to do this.

In [16]:
from sklearn.ensemble import RandomForestClassifier

# Merge the calculated differences with the main dataset
df = df.merge(diff, on='id', how='left')

# Define features and target
X = df.drop(columns_to_drop, axis=1)
y = df['churn']

In [17]:
df.head()

id                     channel_sales   
0  24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua  \
1  d29c2c54acc38ff3c0614d0a653813dd                           MISSING   
2  764c75f661154dac3a6c254cd082ea7d  foosdfpfkusacimwkcsosbicdxkicaua   
3  bba03439a292a1e166f80264c16191cb  lmkebamcaaclubfxadlmueccxoimlema   
4  149d57cf92fc41cf94415803a877cb4b                           MISSING   

   cons_12m  cons_gas_12m  cons_last_month date_activ   date_end   
0         0         54946                0 2013-06-15 2016-06-15  \
1      4660             0                0 2009-08-21 2016-08-30   
2       544             0                0 2010-04-16 2016-04-16   
3      1584             0                0 2010-03-30 2016-03-30   
4      4425             0              526 2010-01-13 2016-03-07   

  date_modif_prod date_renewal  forecast_cons_12m  forecast_cons_year   
0      2015-11-01   2015-06-23               0.00                   0  \
1      2009-08-21   2015-08-31             189.95                   0   
2      2010-04-16   2015-04-17              47.96                   0   
3      2010-03-30   2015-03-31             240.04                   0   
4      2010-01-13   2015-03-09             445.75                 526   

   forecast_discount_energy  forecast_meter_rent_12m   
0                       0.0                     1.78  \
1                       0.0                    16.27   
2                       0.0                    38.72   
3                       0.0                    19.83   
4                       0.0                   131.73   

   forecast_price_energy_off_peak  forecast_price_energy_peak   
0                        0.114481                    0.098142  \
1                        0.145711                    0.000000   
2                        0.165794                    0.087899   
3                        0.146694                    0.000000   
4                        0.116900                    0.100015   

   forecast_price_pow_off_peak  has_gas  imp_cons  margin_gross_pow_ele   
0                    40.606701        1      0.00                 25.44  \
1                    44.311378        0      0.00                 16.38   
2                    44.311378        0      0.00                 28.60   
3                    44.311378        0      0.00                 30.22   
4                    40.606701        0     52.32                 44.91   

   margin_net_pow_ele  nb_prod_act  net_margin  num_years_antig   
0               25.44            2      678.99                3  \
1               16.38            1       18.89                6   
2               28.60            1        6.60                6   
3               30.22            1       25.46                6   
4               44.91            1       47.98                6   

                          origin_up  pow_max                   price_date_x   
0  lxidpiddsbxsbosboudacockeimpuepw   43.648  2015-06-16 12:00:00.000000000  \
1  kamkkxfxxuwbdslkwifmmcsiusiuosws   13.800  2015-06-16 12:00:00.000000000   
2  kamkkxfxxuwbdslkwifmmcsiusiuosws   13.856  2015-06-16 12:00:00.000000000   
3  kamkkxfxxuwbdslkwifmmcsiusiuosws   13.200  2015-06-16 12:00:00.000000000   
4  kamkkxfxxuwbdslkwifmmcsiusiuosws   19.800  2015-06-16 12:00:00.000000000   

   mean_year_price_p1_var  mean_year_price_p2_var  mean_year_price_p3_var   
0                0.124787                0.100749                0.066530  \
1                0.149609                0.007124                0.000000   
2                0.170512                0.088421                0.000000   
3                0.151210                0.000000                0.000000   
4                0.124174                0.103638                0.072865   

   mean_year_price_p1_fix  mean_year_price_p2_fix  mean_year_price_p3_fix   
0               40.942265               22.352010               14.901340  \
1               44.311375                0.000000                0.00000

In [18]:
# Initialize Random Forest Classifier
rf = RandomForestClassifier(random_state=100)

# Train the model
rf.fit(X, y)

# Get feature importances
importances = rf.feature_importances_

# Create a DataFrame for feature importances
feature_importances = pd.DataFrame({'feature': list(X.columns), 'importance': importances}).sort_values('importance', ascending = False)

# Display feature importances
print('Feature Importances:')
print(feature_importances.reset_index(drop=True))

# Define a threshold
threshold = 0.01

# Select features with importance above the threshold
selected_features = feature_importances[feature_importances['importance'] > threshold]['feature']

# Display selected features
print('\nSelected Features:')
print(selected_features.reset_index(drop=True))

Feature Importances:
                           feature  importance
0                         cons_12m    0.063812
1          forecast_meter_rent_12m    0.062524
2                       net_margin    0.061332
3                forecast_cons_12m    0.058228
4             margin_gross_pow_ele    0.054013
5               margin_net_pow_ele    0.052114
6                          pow_max    0.043023
7                  cons_last_month    0.042209
8           mean_year_price_p1_var    0.038404
9                         imp_cons    0.035808
10              forecast_cons_year    0.035436
11              mean_year_price_p1    0.035408
12     offpeak_diff_dec_jan_energy    0.033165
13            mean_6m_price_p1_var    0.029427
14                mean_6m_price_p1    0.026591
15                 num_years_antig    0.024162
16  forecast_price_energy_off_peak    0.021948
17                mean_3m_price_p1    0.021365
18            mean_3m_price_p1_var    0.021336
19          mean_year_price_p1_fix    0

In [19]:
len(selected_features)

28

---

## 5. Modelling

Now, let's continue with building and evaluating the predictive model using the Random Forest classifier.

    Advantages and Disadvantages of using Random Forest for this use case:

    > Advantages:
        - Handles non-linear relationships well.
        - Handles a mixture of numerical and categorical features.
        - Provides feature importances, helping to understand which features are most influential.
        - Robust to overfitting due to ensemble nature.
        - Suitable for high-dimensional data.

    > Disadvantages:
        - May overfit with noisy data if not properly tuned.
        - Can be computationally expensive.
        - Interpretability might be challenging when dealing with a large number of trees.

After selecting the features, we can now train the Random Forest classifier and evaluate its performance.

In [20]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X[selected_features], y, test_size=0.2, random_state=100)

print('Shape of X_train:', X_train.shape)
print('Shape of X_test :', X_test.shape)
print('Shape of y_train:', y_train.shape)
print('Shape of y_test :', y_test.shape)

Shape of X_train: (11684, 28)
Shape of X_test : (2921, 28)
Shape of y_train: (11684,)
Shape of y_test : (2921,)


In [21]:
# Train the Random Forest classifier
rf = RandomForestClassifier(n_estimators=100, random_state=100)
rf.fit(X_train, y_train)

# Predict churn on the test data
y_pred = rf.predict(X_test)

# Evaluate the model
print('Confusion Matrix:')
print(confusion_matrix(y_test, y_pred))
print('\nClassification Report:')
print(classification_report(y_test, y_pred))
print('\nAccuracy Score:', accuracy_score(y_test, y_pred).round(2))

Confusion Matrix:
[[2622    4]
 [ 270   25]]

Classification Report:
              precision    recall  f1-score   support

           0       0.91      1.00      0.95      2626
           1       0.86      0.08      0.15       295

    accuracy                           0.91      2921
   macro avg       0.88      0.54      0.55      2921
weighted avg       0.90      0.91      0.87      2921


Accuracy Score: 0.91


### Interpretation

Based on the confusion matrix and the classification report, the following interpretations can be made:

    > The confusion matrix shows that:
        - The model correctly predicted 2622 true negatives (actual 0, predicted 0) and 25 true positives (actual 1, predicted 1).
        - The model made 4 false positive predictions (actual 0, predicted 1), and 270 false negatives (actual 1, predicted 0).
        - This implies that the model is more accurate at predicting the negative class (0) compared to the positive class (1).

    > The classification report provides additional metrics to evaluate the model's performance. Precision is the ratio of correctly predicted positive observations to the total predicted positives.
        - The precision for the positive class (1) is high (0.86), indicating that when the model predicts an instance as positive, it is likely to be correct.
        - The recall for the positive class (1) is quite low (0.08), indicating that the model is not good at identifying all actual positive instances.
        - The F1-score for the positive class (1) is quite low (0.15), suggesting that the model is not well balanced between precision and recall for this class.

    > The overall accuracy of the model is 0.91, which means the model correctly predicted 91% of all instances.

    > However, accuracy can be misleading if the classes are imbalanced.
    
    > In this case, the model seems to perform much better on the negative class (0) than the positive class (1), which is reflected in the high accuracy but low recall for the positive class.

### Conclusion

In conclusion, while the model has high accuracy and precision, it struggles with recall, particularly for the positive class. This suggests the model may need improvement to better identify positive instances.